src/task_monitor/tasks/modify_tree.rs

```diff
@@ -0,0 +1,231 @@
+
+        // It is very important to generate that event AFTER transaction is committed to
+        // database. Otherwise, notified task may not see changes as transaction was not
+        // committed yet.
+        if request_sync {
+            sync_tree_notify.notify_one();
+        }
+    }
+}
+
+/// Looks for any pending changes to the tree. Returns true if there were any
+/// changes applied to the tree.
+async fn do_modify_tree(
+    tx: &mut Transaction<'_, Postgres>,
+    batch_deletion_timeout: chrono::Duration,
+    min_batch_deletion_size: usize,
+    tree_state: &TreeState,
+) -> anyhow::Result<bool> {
+    let deletions = get_deletions(
+        tx,
+        batch_deletion_timeout,
+        min_batch_deletion_size,
+        tree_state,
+    )
+    .await?;
+
+    // Deleting identities has precedence over inserting them.
+    if !deletions.is_empty() {
+        run_deletions(tx, tree_state, deletions).await
+    } else {
+        run_insertions(tx, tree_state).await
+    }
+}
+
+pub async fn get_deletions(
+    tx: &mut Transaction<'_, Postgres>,
+    batch_deletion_timeout: chrono::Duration,
+    min_batch_deletion_size: usize,
+    tree_state: &TreeState,
+) -> anyhow::Result<Vec<DeletionEntry>> {
+    let deletions = tx.get_deletions().await?;
+
+    if deletions.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    // If the minimum deletions batch size is not reached and the deletion time
+    // interval has not elapsed then we can skip
+    if deletions.len() < min_batch_deletion_size {
+        let last_deletion_timestamp = tx.get_latest_deletion().await?.timestamp;
+        if Utc::now() - last_deletion_timestamp <= batch_deletion_timeout {
+            return Ok(Vec::new());
+        }
+    }
+
+    // Dedup deletion entries
+    let deletions = deletions.into_iter().collect::<HashSet<DeletionEntry>>();
+    let mut deletions = deletions.into_iter().collect::<Vec<DeletionEntry>>();
+
+    // Check if the deletion batch could potentially create:
+    // - duplicate root on the tree when inserting to identities
+    // - duplicate root on batch
+    // Such situation may happen only when deletions are done from the last inserted leaf in
+    // decreasing order (each next leaf is decreased by 1) - same root for identities, or when
+    // deletions are going to create same tree state - continuous deletions.
+    // To avoid such situation we sort then in ascending order and only check the scenario when
+    // they are continuous ending with last leaf index
+    deletions.sort_by(|d1, d2| d1.leaf_index.cmp(&d2.leaf_index));
+
+    if let Some(last_leaf_index) = tree_state.latest_tree().next_leaf().checked_sub(1) {
+        let indices_are_continuous = deletions
+            .windows(2)
+            .all(|w| w[1].leaf_index == w[0].leaf_index + 1);
+
+        if indices_are_continuous && deletions.last().unwrap().leaf_index == last_leaf_index {
+            warn!(
```

Comment:
I'm a little worried that under certain conditions we'd see this warning many times in 5 second intervals (or more frequent).

I'm wondering if it makes sense to bump the deletion timestamp here to delay it further. Or keep an in memory timer just for this condition

Reply:
sounds good